src/Hosting/Hosting/src/Internal/ConfigureBuilder.cs

```diff
@@ -51,9 +57,25 @@ private void Invoke(object instance, IApplicationBuilder builder)
                                 MethodInfo.DeclaringType.FullName), ex);
                         }
                     }
+
+                    // (ParameterType)args[index]
+                    arguments[index] = Expression.Convert(Expression.ArrayIndex(argsArray, Expression.Constant(index)), parameterInfo.ParameterType);
                 }
-                MethodInfo.Invoke(instance, parameters);
+
+                // We're going to build a dynamic method to invoke the Configure method
+                // void Configure(object instance, object[] args)
+                // {
+                //     ((Startup)instance).Configure((ParameterType)args[1..n])
+                // }
+
+                // ((Startup)instance)
+                var instanceExpr = Expression.Convert(instanceArg, instance.GetType());
+
+                // (Startup)instance).Configure(...)
+                var bodyExpr = Expression.Call(instanceExpr, MethodInfo, arguments);
+                var invoker = Expression.Lambda<Action<object, object[]>>(bodyExpr, instanceArg, argsArray).Compile();
```

Comment:
I can avoid expression compilation in the default template cases:
- Startup.Configure(IApplicationBuilder)
- Startup.Configure(IApplicationBuilder, IHostingEnvironment)